Use combo block to model vessels

JonathanPham opened the pull request:
Addresses Issue #38 

Implemented the `ComboBlock` in `blocks.py` to allow users to model vessels using an arbitrary combination of a resistor, a capacitor, an inductor, and/or a stenosis element. This `ComboBlock` effectively replaces the `R`, `RC`, `RL`, `RCL`, and `StenosisBlock` elements, since we can model all of these using the `ComboBlock` now.

Tested the accuracy of the `ComboBlock` by checking that it yields the same 0d simulation results as a `StenosisBlock` for the attached 0d input files (from VMR). 
[zero_d_json.zip](https://github.com/SimVascular/svZeroDSolver/files/6871814/zero_d_json.zip)

Here's an example of how to use the `ComboBlock` to model three different elements: 1) `R-C-L-stenosis`, 2) `R-L`, and 3) `R-C-stenosis`:

```
"vessels": [
        {
            "boundary_conditions": {
                "outlet": "RCR_2"
            },
            "vessel_id": 4,
            "vessel_length": 5.182877077031685,
            "vessel_name": "branch4_seg0",
            "zero_d_element_type": "COMBO", # an R-C-L-stenosis element
            "zero_d_element_values": {
                "R_poiseuille": 24.134876235052637,
                "C" : 0.0001,
                "L" : 1.0,
                "stenosis_coefficient": 1.2547136459605657 
            }
        },
        {
            "boundary_conditions": {
                "outlet": "RCR_3"
            },
            "vessel_id": 5,
            "vessel_length": 4.060560280870732,
            "vessel_name": "branch5_seg0",
            "zero_d_element_type": "COMBO", # an R-L element
            "zero_d_element_values": {
                "R_poiseuille": 3.0172124611092825,
                "L" : 10.0
            }
        },
        {
            "boundary_conditions": {
                "outlet": "RCR_4"
            },
            "vessel_id": 6,
            "vessel_length": 11.544295833769882,
            "vessel_name": "branch6_seg0",
            "zero_d_element_type": "COMBO", # an R-C-stenosis element
            "zero_d_element_values": {
                "R_poiseuille": 10.256448651856623,
                "C" : 0.0001,
                "stenosis_coefficient": 4.732081168911113 
            }
        }
    ]
```

We can also easily extend the new json input file format and the `ComboBlock` implementation to incorporate more complex features, such as the ones described in **M. Mirramezani and S.C. Shadden. A distributed lumped parameter model of blood flow. Annals of Biomedical Engineering. 2020.**

Also resolving a small part of Issue #37. Added description comments to the `json` input files for the test cases.

mrp089 commented on svzerodsolver/blocks.py:
```diff
@@ -275,6 +275,33 @@ def update_solution(self, args):
+        self.mat['F'] = [(1.0, -1.0 * self.stenosis_coefficient * np.abs(Q_in) - self.R, -1.0, 0), (0, 1.0, 0, -1.0)]
+        self.mat['dF'] = [(0, -1.0 * self.stenosis_coefficient * np.abs(Q_in), 0, 0), (0,) * 4]
+
+
```
The equations look good, but I'm not sure how versatile this is. Can you checek the following?
1. The model block works for all combinations of `R`, `C`, `L`, and `stenosis` (`0` or given in input file) and doesn't result in a singular system (e.g. by introducing a row of zeros)
2. You don't introduce unnecessary `DOFs`

mrp089 replied:
It looks like we now have redundant code: Individual `R`, `C`, `L`, and `stenosis` blocks plus the `Combo` block. Can we either use only `ComboBlock` and delete the individual blocks or retrieve `E`, `F`, and `dF` matrices from the individual blocks?

JonathanPham replied:
1. The ComboBlock does work for all combinations of `R`, `C`, `L`, and `stenosis`. I updated all test cases to use the ComboBlock and all tests pass. Furthermore, since capacitors and inductors are never used by themselves to model vessels and they must instead always be used in conjunction with some kind of resistor instead (which includes a Poiseuille-based resistor), I implemented a short check to ensure that the user provides a Poiseuille resistance value when they use the `ComboBlock`.
2. We don't introduce any unnecessary DoFs. We retain the same set of DoFs, (P_in, Q_in, P_out, Q_in), for the `ComboBlock`. There are not unnecessary "internal variables" in the `ComboBlock`.
3. Good idea. To prevent redundancy and possible confusion in new users, I deleted the individual blocks, e.g. `RC`, `RL`, `Resistance`, etc. Now, the only way to specify these blocks is to use the `ComboBlock`.

mrp089 commented on svzerodsolver/solver.py:
```diff
@@ -320,6 +320,12 @@ def create_vessel_blocks(parameters, custom_0d_elements_arguments):
+            C = vessel_id_to_zero_d_element_dict[vessel_id]["zero_d_element_values"]["C"] if "C" in vessel_id_to_zero_d_element_dict[vessel_id]["zero_d_element_values"] else 0
+            L = vessel_id_to_zero_d_element_dict[vessel_id]["zero_d_element_values"]["L"] if "L" in vessel_id_to_zero_d_element_dict[vessel_id]["zero_d_element_values"] else 0
+            stenosis_coefficient = vessel_id_to_zero_d_element_dict[vessel_id]["zero_d_element_values"]["stenosis_coefficient"] if "stenosis_coefficient" in vessel_id_to_zero_d_element_dict[vessel_id]["zero_d_element_values"] else 0
+            vessel_blocks[block_name] = ntwku.ComboBlock(R = R, C = C, L = L, stenosis_coefficient = stenosis_coefficient, connecting_block_list = connecting_block_list, name = block_name, flow_directions = flow_directions)
```
If we have the `COMBO` block that covers the same elements as individual blocks, should we default to `COMBO` (and remove the other cases since they are redundant)?

JonathanPham replied:
I originally left the other blocks, such as `RL`, `RC`, `RCL`, etc. in `solver.py` because those other blocks were already fully tested and it doesn't necessarily hurt to have both the ComboBlock implementation and the other blocks. But this does create an extra redundancy that could potentially confuse users in the future. I'll go ahead and remove these other blocks from `solver.py` and `blocks.py`.

JonathanPham replied:
Addressed in [this commit](https://github.com/SimVascular/svZeroDSolver/pull/42/commits/5543aa0aadd4ad129616189ced96b29a379e1e7e).

ktbolt commented:
Comments are indeed a good idea. However, there is no need to use an underscore `_comments` field. Why not just have a `description` field?

JonathanPham commented on test_cases/steadyFlow_RC_R.json:
```diff
@@ -1,4 +1,7 @@
 {
+    "_comments": {
+            "description of test case" : "steady flow -> RC -> R"
+    },
```
Good idea, I'll add in a `description` field (as per [Dave's comment](https://github.com/SimVascular/svZeroDSolver/pull/42#issuecomment-886902731)) to provide a brief overview of how we compute the analytical results.

JonathanPham commented:
Capacitors or inductors are never used **by themselves** to model blood vessels. Instead, they are used in conjunction with some kind of resistor, e.g. a Poiseuille resistor. Thus, I also removed the individual capacitor and inductor blocks in `solver.py` and `blocks.py` to prevent novice misuse of these blocks.

mrp089 commented:
Awesome, @JonathanPham, thank you for these quick changes! Do you think it makes sense to get rid of the `zero_d_element_type` entirely and just default to `ComboBlock` (with whatever `zero_d_element_values` are given)?

Let me know if you have any objections! I'm mainly thinking of "classic" cardiovascular models, but you have a broader view of 0D capabilities.

mrp089 commented:
If we go through with that we could rename `Combo` to `BloodVessel` or something similar.

JonathanPham commented:
@mrp089 Good question. While the `BloodVessel` (originally `ComboBlock`) block is now the only built-in block capable of modleing blood vessels, there are still some advanced users (e.g. Ingrid and Mattia) who want to develop their own blocks for modeling vessels. As such, I think we should leave the `zero_d_element_type` field in the `json` input file so that those advanced users can use their own special blocks.

JonathanPham commented:
@mrp089 and @ktbolt Thanks for your suggestions!

The following updates were made:

1. Deleted the redundant blocks (`R`, `RC`, `RL`, `RCL`, `Stenosis`) that could be modeled using the `ComboBlock`. I also updated the all tests cases reflect these changes. All test cases now use the `ComboBlock`.
2. Capacitors or inductors are never used **by themselves** to model blood vessels. Instead, they are used in conjunction with some kind of resistor, e.g. a Poiseuille resistor. Thus, I also removed the individual capacitor and inductor blocks in `solver.py` and `blocks.py` to prevent novice misuse of these blocks.
3. Furthermore, to prevent specification of only capacitance or inductance in the `ComboBlock`, we need to require that users provide a viscous Poiseuille-based resistance when they use the `ComboBlock`. This requirement also aligns with our previous `RC`, `RL`, `RCL`, and `StenosisBlock` implementations, all of which required specification of a viscous Poiseuille-based resistance.
4. Added description and derivation of analytical results for each test case
5. Renamed `ComboBlock` to `BloodVessel` since the `ComboBlock` is now the only (built-in) block capable of modeling blood vessels.
